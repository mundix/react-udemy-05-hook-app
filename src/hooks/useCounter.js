// custom hooks no es mas que una funcion 
import { useState } from "react";

export const useCounter = ( initialState = 10) => {
    
    const [counter, setCounter] = useState(initialState);

    const increment = (factor = 1) => {
        setCounter(counter + factor);
    }

    const decrement = (factor = 1) => {
        setCounter(counter - factor);
    }

    const reset = () => {
        setCounter(initialState);
    }
    // se puede regresar un objeto { } o un array dependiendo la situacion
    //este custom hook esta extrayendo la logica de mi contador 
    return {
        counter,
        increment,
        decrement,
        reset
    }
}