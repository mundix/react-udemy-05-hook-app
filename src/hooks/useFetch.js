import { useEffect, useRef, useState } from "react"

export const useFetch = (url) => {

    const isMounted = useRef(true);
    const [state, setState] = useState({ data: null, loading: true, error: null });

    useEffect(()=>{
        return ()=>{
            isMounted.current = false;
        }
    }, []);


    useEffect(() => {
        setState({ data: null, loading: true, error: null }); //reinicio los valores para qeu salga el loading
        fetch(url)
            .then(resp => resp.json())
            .then(data => {
                // setTimeout(() => {
                    if(isMounted.current) {
                        setState({
                            loading: false,
                            error: null,
                            data
                        });    
                    } else {
                        // console.log('setState no se llamo');
                    }
                // }, 4000);

            })
            .catch(() =>{
                setState({
                    data: null,
                    loading: false,
                    error: 'No se puso cargar la info'
                })
            });
    }, [url]);

    //solo regresamos el state
    return state;
}
