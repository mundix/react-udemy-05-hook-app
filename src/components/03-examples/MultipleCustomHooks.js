import React from 'react';
import { useFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';
import '../02-useEffect/effects.css';


export const MultipleCustomHooks = () => {
    const {counter, increment} = useCounter(1);
    const url = `https://www.breakingbadapi.com/api/quotes/${counter}`;
    const { loading, data } = useFetch(url);
    
    const {author, quote } = !!data && data[0]; //el !!data si lelga la data asigna el data en la posicion 0
    

    return (
        <div>
            <h1>Braking Bad Quotes {JSON.stringify(loading, null, 3)}</h1>
            <hr />

            {loading ?
                (
                    <div className="alert alert-info text-center">
                        Loading
                    </div>
                ) :
                (
                    <blockquote className="blockquote text-right">
                        <p className="mb-0">{quote}</p>
                        <footer className="blockquote-footer">{author}</footer>
                    </blockquote>
                )
            }

            <button className="btn btn-primary"
                onClick={()=>increment(1)}
            >
                Siguiente Quote
            </button>



        </div>
    )
}
