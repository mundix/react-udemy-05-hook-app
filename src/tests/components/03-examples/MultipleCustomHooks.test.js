import React from 'react';
import { shallow } from 'enzyme';
// Compoennts 
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
// Custo hooks
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');


describe('Pruebas en MultipleCustomHooks', () => {
    // El useCounter fuera de los test, para usarlo en cualquier lado. 
    // Regresa el objeto pero solo usaremor counter y el increment
    // {
    //      counter,
    //      increment,
    //      decrement,
    //      reset
    // }
    useCounter.mockReturnValue({
        counter: 10,
        increment: ()=>{}
    })

    test('Muestra Correctmente MultipleCustomHooks', () => {
        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks />);
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe mostrar la informacion', () => {
        // mockReturnValue se usa para pasar argumentos de pruebas (o sea Falsos),
        // pero este valor es el que usa  el useState dentro useFetch ? 
        // Si el argumento de useFetch es el url 
        useFetch.mockReturnValue({
            data: [{
                author: 'Edmundo',
                quote: 'Hola Mundo',
            }],
            loading: false,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks />);
        expect(wrapper.find('.alert').exists()).toBe(false); //No debe existir si tiene informacion
        // wrapper.find('el').text().trim()  -> limpia el texto de esos elementos.
        expect(wrapper.find('.mb-0').text().trim()).toBe('Hola Mundo'); //Verifica si teine l texto Hola mundo el quote
        expect(wrapper.find('footer').text().trim()).toBe('Edmundo'); // Verifica el nombre del author 
        // console.log(wrapper.html());
    })


})
