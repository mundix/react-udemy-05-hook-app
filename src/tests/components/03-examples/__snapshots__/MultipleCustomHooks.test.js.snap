// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en MultipleCustomHooks Muestra Correctmente MultipleCustomHooks 1`] = `
<div>
  <h1>
    Braking Bad Quotes 
    true
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente Quote
  </button>
</div>
`;
